sync/src/chain.rs

```diff
@@ -328,117 +335,125 @@ type RlpResponseResult = Result<Option<(PacketId, RlpStream)>, PacketDecodeError
+		self.new_blocks.reset();
 		self.snapshot.clear();
 		if self.state == SyncState::SnapshotData {
 			debug!(target:"sync", "Aborting snapshot restore");
 			io.snapshot_service().abort_restore();
 		}
 		for (_, ref mut p) in &mut self.peers {
-			p.asking_blocks.clear();
-			p.asking_hash = None;
-			// mark any pending requests as expired
-			if p.asking != PeerAsking::Nothing && p.is_allowed() {
-				p.expired = true;
+			if p.block_set != Some(BlockSet::OldBlocks) {
+				p.asking_blocks.clear();
+				p.asking_hash = None;
+				// mark any pending requests as expired
+				if p.asking != PeerAsking::Nothing && p.is_allowed() {
+					p.expired = true;
+				}
 			}
 		}
-		self.syncing_difficulty = From::from(0u64);
 		self.state = SyncState::Idle;
 		self.active_peers = self.peers.keys().cloned().collect();
 	}
 
 	/// Restart sync
-	pub fn restart(&mut self, io: &mut SyncIo) {
+	pub fn reset_and_continue(&mut self, io: &mut SyncIo) {
 		trace!(target: "sync", "Restarting");
 		self.reset(io);
-		self.start_sync_round(io);
 		self.continue_sync(io);
 	}
 
 	/// Remove peer from active peer set
-	fn deactivate_peer(&mut self, io: &mut SyncIo, peer_id: PeerId) {
+	fn deactivate_peer(&mut self, _io: &mut SyncIo, peer_id: PeerId) {
 		trace!(target: "sync", "Deactivating peer {}", peer_id);
 		self.active_peers.remove(&peer_id);
-		if self.active_peers.is_empty() {
-			trace!(target: "sync", "No more active peers");
-			if self.state == SyncState::ChainHead {
-				self.complete_sync(io);
-			} else {
-				self.restart(io);
-			}
-		}
 	}
 
 	fn start_snapshot_sync(&mut self, io: &mut SyncIo, peer_id: PeerId) {
 		self.snapshot.clear();
 		self.request_snapshot_manifest(io, peer_id);
 		self.state = SyncState::SnapshotManifest;
 	}
+
+	/// Restart sync disregarding the block queue status. May end up re-downloading up to QUEUE_SIZE blocks
+	pub fn restart(&mut self, io: &mut SyncIo) {
+		self.init_downloaders(io.chain());
+		self.reset_and_continue(io);
+	}
   
 	/// Restart sync after bad block has been detected. May end up re-downloading up to QUEUE_SIZE blocks
-	fn restart_on_bad_block(&mut self, io: &mut SyncIo) {
+	fn init_downloaders(&mut self, chain: &BlockChainClient) {
 		// Do not assume that the block queue/chain still has our last_imported_block
-		let chain = io.chain().chain_info();
-		self.last_imported_block = chain.best_block_number;
-		self.last_imported_hash = chain.best_block_hash;
-		self.restart(io);
+		let chain = chain.chain_info();
+		self.new_blocks = BlockDownloader::new(false, &chain.best_block_hash, chain.best_block_number);
+		if let (Some(ancient_block_hash), Some(ancient_block_number)) = (chain.ancient_block_hash, chain.ancient_block_number) {
+
+			trace!(target: "sync", "Downloading old blocks from {:?} (#{}) till {:?} (#{:?})", ancient_block_hash, ancient_block_number, chain.first_block_hash, chain.first_block_number);
+			let mut downloader = BlockDownloader::new(true, &ancient_block_hash, ancient_block_number);
```

Comment:
I agree that it is better to download the ancient chain in reverse, but will leave it for a future PR